src/julia_gc.c

```diff
@@ -0,0 +1,755 @@
+
+    // mark all global objects
+    for (Int i = 0; i < GlobalCount; i++) {
+        Bag p = *GlobalAddr[i];
+        if (IS_BAG_REF(p)) {
+            JMark(cache, sp, p);
+        }
+    }
+
+    // scan the GAP call stack, too
+    // FIXME: is this really necessary? STATE(CurrLVars) is already marked as
+    // a global object (via GlobalAddr above), and it is the head of a linked
+    // list containing the others, so it should not be necessary (and a quick
+    // test confirms this
+    MarkStackFrames(cache, sp, STATE(CurrLVars));
+    for (Bag execState = CurrExecState(); execState;
+         execState = ELM_PLIST(execState, 1)) {
+        MarkStackFrames(cache, sp, ELM_PLIST(execState, 2));
+    }
+}
+
+// helper function to test if Julia considers an object to
+// be in the old generation
+static inline int GcOld(void * p)
+{
+    return (jl_astaggedvalue(p)->bits.gc & 2) != 0;
+}
+
+// the Julia marking function for master pointer objects (i.e., this function
+// is called by the Julia GC whenever it marks a GAP master pointer object)
+static void JMarkMPtr(void * cache, void * sp, void * obj)
+{
+    if (!*(void **)obj)
+        return;
+    if (JMark(cache, sp, BAG_HEADER(obj)) && GcOld(obj))
+        jl_gc_mark_push_remset(JuliaTLS, obj, 1);
+}
+
+// the Julia marking function for bags (i.e., this function is called by the
+// Julia GC whenever it marks a GAP bag object)
+static void JMarkBag(void * cache, void * sp, void * obj)
+{
+    JCache = cache;
+    JSp = sp;
+    BagHeader * hdr = (BagHeader *)obj;
+    Bag         contents = (Bag)(hdr + 1);
+    UInt        tnum = hdr->type;
+    YoungRef = 0;
+    OldObj = GcOld(obj);
+    TabMarkFuncBags[tnum]((Bag)&contents);
+    if (OldObj && YoungRef)
+        jl_gc_mark_push_remset(JuliaTLS, obj, YoungRef);
+}
+
+
+void InitBags(UInt              initial_size,
+              TNumStackFuncBags stack_func,
+              Bag *             stack_bottom,
+              UInt              stack_align)
+{
+    // HOOK: initialization happens here.
+    jl_root_scanner_hook = GapRootScanner;
+    jl_gc_disable_generational = 0;
+    jl_nonpool_alloc_hook = alloc_bigval;
+    jl_nonpool_free_hook = free_bigval;
+    for (UInt i = 0; i < NTYPES; i++)
+        TabMarkFuncBags[i] = MarkAllSubBags;
+    jl_extend_init();
+    JuliaTLS = jl_extend_get_ptls_states();
+    // jl_gc_enable(0); /// DEBUGGING
+    max_pool_obj_size = jl_extend_gc_max_pool_obj_size();
+    Module = jl_new_module(jl_symbol("ForeignGAP"));
+    datatype_mptr = jl_new_foreign_type(jl_symbol("Bag"), Module, jl_any_type,
+                                        JMarkMPtr, NULL, 1, 0);
+    datatype_bag = jl_new_foreign_type(jl_symbol("BagInner"), Module,
+                                       jl_any_type, JMarkBag, NULL, 1, 0);
+    datatype_largebag = jl_new_foreign_type(
+        jl_symbol("BagInner"), Module, jl_any_type, JMarkBag, NULL, 1, 1);
```

Comment:
Hmm, so two datatypes both use `BagInner` as their name. Perhaps this second one should be `BagLargeInner`?

Or perhaps it should be `MasterPointer` (or `MPtr`), `Bag`, `LargeBag`, matching the C variable names?